### ISSUE TITLE
ByteToMessageDecoder behaves gracelessly when channelInactive is triggered while decode() is on the stack.

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#108]
### Expected behavior

When a `ByteToMessageDecoder` calls `ctx.close()` during its processing, and that `close` completes synchronously, receiving `channelInactive` will not cause ByteToMessageDecoder to re-parse previously parsed data.

### Actual behavior

When this happens, `ByteToMessageDecoder.channelInactive` calls `decodeLast` with a fresh copy `self.cumulationBuffer`. By default, `decodeLast` is the same as `decode`, so this essentially calls `decode` in almost all cases. This means that the call stack ends up looking like this:

1. `decode()`
2. `ctx.close()`
...
3. `decodeLast()`
4. `decode()`

Worse, this second call to `decode` does it with an out of date copy of the `cumulationBuffer`, which it necessarily must do because we always operate on a copy of the `cumulationBuffer` and save it off after our calls to `decode`.

### If possible, minimal yet complete reproducer code (or URL to code)

```swift
final class BufferLengthReporter: ByteToMessageDecoder {
    typealias InboundIn = ByteBuffer
    typealias InboundOut = Int

    var cumulationBuffer: ByteBuffer?

    func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
        ctx.fireChannelRead(self.wrapInboundOut(buffer.readableBytes))
        buffer.moveReaderIndex(forwardBy: 4)
        ctx.close(promise: nil)
        return .needMoreData
    }

    func decodeLast(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
        ctx.fireChannelRead(self.wrapInboundOut(buffer.readableBytes))
        return .needMoreData
    }
}

class Test: XCTestCase {
    func testDontRepeatedlyParse() throws {
        let channel = EmbeddedChannel()
        defer {
            _ = try? channel.finish()
        }

        XCTAssertNoThrow(try channel.pipeline.add(handler: BufferLengthReporter()).wait())

        var buffer = channel.allocator.buffer(capacity: 5)
        buffer.write(staticString: "hello")

        XCTAssertTrue(try channel.writeInbound(buffer))
        XCTAssertFalse(channel.isActive)

        XCTAssertEqual(5, channel.readInbound())
        XCTAssertEqual(1, channel.readInbound())
        XCTAssertEqual(nil as Int?, channel.readInbound())
    }
}
```

The second `XCTAssertEqual` in the final block fails, reporting `5` instead of `1`.

### SwiftNIO version/commit hash

1.1.0

#### 1. Remove `decodeLast`.

`decodeLast` is a strange function. It exists for protocols for whom EOF is a meaningful signal, but those protocols are not that common. More importantly, in the case of those protocols they could simply implement `channelInactive` themselves to handle the case. Right now, for the default protocol case, `decodeLast` is simply a waste of CPU: the protocol couldn't decode the buffer before, no reason to assume it can now.

The advantage of removing `decodeLast` is that it becomes clear that this is a thing users need to handle themselves. The disadvantage is that it doesn't solve this problem for handlers that treat EOF as meaningful: they still need a way to work out whether they're synchronously receiving EOF from the remote peer.

#### 2. Change the default behaviour of `decodeLast`.

A smaller-scale version of (1), this would just change `decodeLast` to be a no-op. This avoids the current problem, but still provides the hook. The downside of this is, again, we end up letting users fall back into this trap.

#### 3. Manage state better in `ByteToMessageDecoder`.

This is probably my preferred option, but it makes B2MD substantially more complex. Ultimately the core issue is that B2MD will *eagerly* dispatch `channelInactive`, which means that it's partway through parsing when that message is processed. That's just a bad idea. So we could add state to `B2MD` to record that it has seen `channelInactive`, but to not dispatch it while we're inside `decode`. Instead, we'd let `decode` return and then spot the change, and transition to calling `decodeLast`.

This is likely to behave better in the long-term.

#### 4. Prevent `channelInactive` firing synchronously.

This is not a great fix, but most users do not realise that the moment they call out to third-party code they are at risk of being re-entrantly called by all kinds of events. We could, in principle, prevent `close` re-entrantly firing `channelInactive`, but frankly I don't think that's a very good idea. I just wanted to note it here.

All-in-all I think I prefer (3), but I'd like to hear what others think.

**Comment by normanmaurer** [comment]
Yep I think 3) sounds like the best idea here.

**Comment by weissi** [comment]
fixed with #675
